File: depr/camera/async_reader.py
from threading import Thread
from threading import Lock
from frame_counter import WeightedFramerateCounter
import cv2

class CameraReaderAsync:
    def __init__(self, videoSource):
        self.__lock = Lock()
        self.__source = videoSource
        self.Start()
        
    def __ReadAsync(self):
        while True:
            if self.__stopRequested:
                return
            validFrame, frame = self.__source.read()
            if validFrame:
                try:
                    self.__lock.acquire()
                    self.fps.tick()
                    self.__frame = frame
                    self.__lastFrameRead = False
                finally:
                    self.__lock.release()

    def Start(self):
        self.__lastFrameRead = False
        self.__frame = None
        self.__stopRequested = False
        self.__validFrame = False
        self.fps = WeightedFramerateCounter()
        Thread(target=self.__ReadAsync).start()
        
    def Stop(self):
        self.__stopRequested = True

    # Return a frame if we have a new frame since this was last called.
    # If there is no frame or if the frame is not new, return None.
    def Read(self):
        try:
            self.__lock.acquire()
            if not self.__lastFrameRead:
                frame = self.__frame
                self.__lastFrameRead = True
                return frame

            return None
        finally:
            self.__lock.release()

    # Return the last frame read even if it has been retrieved before.
    # Will return None if we never read a valid frame from the source.
    def ReadLastFrame(self):
        return self.__frame

class Camera():
    cvcamera = None
    cvreader = None
    width = 0
    height = 0
    panPos = 0
    tiltPos = 0
    zoomPos = -1
    _badPTZcount = 0
    
    def __init__(self, usbdevnum, width, height):
        # Start by establishing control connection
        #pysca.connect(cfg['socket'])

        # Open video stream as CV camera
        self.cvcamera = cv2.VideoCapture(usbdevnum)
        self.width = width
        self.height = height
        self.cvcamera.set(3, width)
        self.cvcamera.set(4, height)

        self.cvreader = CameraReaderAsync(self.cvcamera)
    
    def lostPTZfeed(self):
        return True
        # return False if self._badPTZcount < 5 else True
            
    def updatePTZ(self):
        self._badPTZcount += 1

File: depr/camera/c_plus_plus/async_reader.cpp
#include "async_reader.h"
#include "frame_counter.h"

namespace Capstone {

    CameraReaderAsync::CameraReaderAsync (const int usbdevnum, const int width, const int height) : cvCamera(usbdevnum) {
        if (!cvCamera.isOpened()) {
            std::cout << "Problem with Camera USB device location: " << usbdevnum << std::endl;
            exit(1);
        }
        cvCamera.set(3, width);
        cvCamera.set(4, height);
        lastFrameRead = false;
        stopRequested = false;
        validFrame = false;
        th.push_back(std::thread(&CameraReaderAsync::readAsync, this));
    }

    void CameraReaderAsync::readAsync (){
        while (true) {
            if (stopRequested) { return; }
            cv::Mat tempframe;
            validFrame = cvCamera.read(tempframe);
            if (validFrame) {
                std::lock_guard<std::mutex> lk(mtx);
                fps.tick();
                frame = tempframe.clone();
                lastFrameRead = false;
            }

        }
    }

    void CameraReaderAsync::stop() {
        stopRequested = true;
    }

    void CameraReaderAsync::release() {
        cvCamera.release();
    }

    cv::Mat CameraReaderAsync::read() {
        std::lock_guard<std::mutex> lk(mtx);
        if (!lastFrameRead) {
            lastFrameRead = true;
            return frame;
        }
        return cv::Mat();
    }

    cv::Mat CameraReaderAsync::readLastFrame() {
        return frame;
    }

    Camera::Camera (const int usbdevnum, const double width, const double height)
    : CameraReaderAsync(usbdevnum, width, height) {
        panPos = 0;
        tiltPos = 0;
        zoomPos = 0;
        badCountPTZ = 0;
    };

    bool Camera::lostPTZfeed() {
        if (badCountPTZ < 5) { return false; }
        else { return true; }
    }

    void Camera::updatePTZ() {
        badCountPTZ++;
    }
}
